### DISCUSSION TITLE
Null support for eq filter

### DISCUSSION THREAD

**Discussion by TimurKr** [PostgREST/postgrest#2893] [Q&A] [answered]
Hi, in SQL when filtering based on an equality, the `=` symbol is used, but if the valzne is null, the `is` filter must be used. I am quite new to SQL and I have more background in programming, where almost always when you are checking equality of two values, whether any of them is null or not, the `==` symbol works. The same obviously applies for `neq` and `!=`.

It is my view of this library, that it should serve as a bridge between the SQL and whatever frontend the developer uses. Therefore I think that this gap between behaviors would be best closed here. I would suggest adding a check in the `eq` and `neq` filters, where if the value is null, the `is` filter is used instead of eq. 

I would like anyone's opinions, as I first tried changing this behavior in a [supabase-flutter](https://github.com/supabase/supabase-flutter/pull/580) library, where I was told it was a simple wrapper around PostgREST, so I thought about asking here.

Also I would have opened a pull request, but I have zero experience with Haskell, or an issue, but this isn't so much of an issue, rather than me being curious if anyone thinks the same. 

Thanks for any opinions!

**Comment by laurenceisla** [comment, marked as answer]
Hey. If I'm understanding correctly, your suggestion is that PostgREST should detect an `eq.null`  and translate it to `is.null`, right? Something like this:

```
GET /people?name=eq.null&age=neq.null
```

Into this:

```
GET /people?name=is.null&age=not.is.null
```

I understand your POV but I'm not sure if PostgREST should be that opinionated in this case. The thing is that DBAs using the tool could argue the opposite: "why is `name=eq.null` filtering the null values when `name = null` should return no rows at all?" So, my opinion is that PostgREST should be consistent when mapping operators with PostgreSQL (not to the front end), that is: `eq` will always be translated to `=`. 

But I think it still in debate if we should use `null` (without quotes) not as a string (which is done right now) but as a `NULL` value in SQL. Related to #2879.

**Reply by TimurKr** [reply]
Hey! Thanks for the reply, I understand. Do you have any insights into why `=` doesn't work with null values in SQL or what are any benefits of that behavior, just out of curiosity. I always thought about it as the same operation as comparing two non-null values, but SQL uses different operators. Thanks!

**Reply by laurenceisla** [reply]
Thanks for the question! PostgreSQL follows the SQL standard, and explains why it behaves this way in [their docs](https://www.postgresql.org/docs/15/functions-comparison.html):

> Do not write expression = NULL because NULL is not “equal to” NULL. (The null value represents an unknown value, and it is not known whether two unknown values are equal.)

And, interestingly, there's a tip that mentions this:

> Some applications might expect that expression = NULL returns true if expression evaluates to the null value. It is highly recommended that these applications be modified to comply with the SQL standard. However, if that cannot be done the [transform_null_equals](https://www.postgresql.org/docs/15/runtime-config-compatible.html#GUC-TRANSFORM-NULL-EQUALS) configuration variable is available. If it is enabled, PostgreSQL will convert x = NULL clauses to x IS NULL.

So there's an option in PostgreSQL that does what you said in the OP. But I understand that it's more an exception for certain apps and should not be the first choice (according to [the entry about that option](https://www.postgresql.org/docs/15/runtime-config-compatible.html#GUC-TRANSFORM-NULL-EQUALS)).

**Reply by TimurKr** [reply]
Thank you so much for pointing me in the right direction. It makes sense now. Massive help!

**Reply by TimurKr** [reply]
Hi @laurenceisla,
sorry to bother you with this issue again. I modified the **transform_null_equals** setting to on, and now when I am querying directly using the SQL, the  `= null` works without an issue. Thanks for pointing this out to me. I wish I didn't have to, but this seems like my only reasonable fix for my issue right now.

But I am using a frontend library (supabase-flutter), which internaly uses PostgREST. Here though,  the `eq.null` is still returning a PostgREST exception. I tried good old curl commands as well, and it isn't working either. Isn't there somewhere in PostgREST a check for validity of the datatype even before the query is run? Or do you have any other ideas why this might not be working?

Thank you!

**Reply by laurenceisla** [reply]
Hi. Yes, the result you're getting is expected; now I notice that I didn't said it clearly in my answer. The thing is PostgREST reads `column=eq.null` and transforms it to `column = 'null'`, notice the `'`. So It's comparing it to a string not the `null` value, which is the expected behavior. In summary, right now there's no direct solution for what you're requesting.

Some alternatives are to make that change (`eq.null` to `is.null`) in your frontend app, or maybe use Nginx and do the parameter change there (it could be adapted from [this example](https://stackoverflow.com/a/76378484/17265105) on a different issue). I'm not sure if or how the last one can be done in Supabase though.

**Reply by TimurKr** [reply]
Thanks for the clarification, that makes perfect sense. I'll try to find a workaround in the frontend.
